HPCC-25630 Metrics milestone 2 - Add a logging sink

kenrowland opened the pull request:
Added metrics sink to write metric collection reports to the log

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [ ] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-25630 Metrics milestone 2 - Adding a logging sink

Added metrics sink to write metric collection reports to the log

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

HPCCSmoketest commented:
Process of PR-14777, label: HPCC-25630 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 07d655c674ddd5016cf0ed0e7888bdeb37964ecb
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-204.ca-central-1.compute.internal

kenrowland commented:
@ghalliday @rpastrana Please review.

Few notes:

- Created new periodic sink base class to handle both log and file sinks since they operate the same way
- Removed 'sink' from the sink type when specifying via settings. Now added automatically (this means the _type_ setting value is now _log_ or _file_ instead of _logsink_ or _filesink_ (more intuitive since they are all sinks.
- Some cleanup of unused header references in the metrics lib.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-204.ca-central-1.compute.internal
Sha: 07d655c674ddd5016cf0ed0e7888bdeb37964ecb
Build: success
Milestone:Install hpccsystems-platform-community_8.0.1-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |158 |158 |0 |0 |0 |58 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |0 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |6 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |26 sec  (00:00:26) |
|setup (thor) |9 |9 |0 |41 sec  (00:00:41) |
|setup (roxie) |19 |19 |0 |18 sec  (00:00:18) |
|test (hthor) |963 |963 |0 |785 sec  (00:13:05) |
|test (thor) |875 |875 |0 |1120 sec  (00:18:40) |
|test (roxie) |1036 |1036 |0 |879 sec  (00:14:39) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  606 sec (00:10:06) |  117 sec (00:01:57) |   21 sec (00:00:21) |   17 sec (00:00:17) | 1435 sec (00:23:55) |   19 sec (00:00:19) | 2224 sec (00:37:04) |

rpastrana commented on system/metrics/testing/metrictest.cpp:
```diff
@@ -64,13 +64,26 @@ const char *testConfigYml = R"!!(component:
     //
     // Simulate retrieving the component and global config
-    Owned<IPropertyTree> pSettings = createPTreeFromYAMLString(testConfigYml, ipt_none, ptr_ignoreWhiteSpace, nullptr);
+    Owned<IPropertyTree> pSettings = createPTreeFromYAMLString(testFileSinkConfigYml, ipt_none, ptr_ignoreWhiteSpace, nullptr);
```
couldn't tell, but is this only built/packaged in DEBUG mode?

rpastrana replied:
Conditionally exclude the entire testing subproject based on "${CMAKE_BUILD_TYPE}"

ghalliday replied:
I don't think it is a problem for testing code to be built in release mode - it is often useful.

ghalliday replied:
(provided it does not significantly increase the size of the deployed system.)  This is separate so it wouldn't.

rpastrana replied:
@ghalliday how do you envision this being used in "release"? I can see perhaps as part of smoke test or regression tests, but I still contend this and code like it shouldn't be packaged in our official release akin to Java junit tests which are built with the project, executed if desired, but not part of the target package.

ghalliday replied:
Well it should be using the cppunit framework that is used by all the other unit tests.  (Which is deployed on a separate option.)  The unit tests are often run in release as well as debug mode.  I agree it shouldn't be deployed as part of our standard release.

rpastrana reviewed (commented):
@kenrowland a few comments

ghalliday commented on system/jlib/jlog.hpp:
```diff
@@ -845,6 +845,7 @@ constexpr LogMsgCategory MCdebugInfo(MSGAUD_programmer, MSGCLS_information, Debu
 constexpr LogMsgCategory MCauditInfo(MSGAUD_audit, MSGCLS_information, AudMsgThreshold);
 constexpr LogMsgCategory MCstats(MSGAUD_operator, MSGCLS_progress, ProgressMsgThreshold);
 constexpr LogMsgCategory MCoperatorInfo(MSGAUD_operator, MSGCLS_information, InfoMsgThreshold);
+constexpr LogMsgCategory MCmetrics(MSGAUD_operator, MSGCLS_all, CriticalMsgThreshold);
```
MSGCLS_all is used for filters, not for logging.  I think there should probably be a new MSGCLS_metric, which should be used here.  I will look at adding one.

ghalliday replied:
The change has been merged to 8.0.x  please rebase and use the new class.

ghalliday commented on system/metrics/sinks/file/CMakeLists.txt:
```diff
@@ -18,11 +18,13 @@ project(hpccmetrics_filesink)
 
 set ( srcs
         fileSink.cpp
+        ${HPCC_SOURCE_DIR}/system/metrics/sinks/common/periodicSink.cpp
```
We try and avoid including the same source file in multiple places.

ghalliday commented on system/metrics/sinks/common/periodicSink.cpp:
```diff
@@ -0,0 +1,83 @@
+/*##############################################################################
+    HPCC SYSTEMS software Copyright (C) 2021 HPCC Systems®.
```
This makes sense to be commoned up - but should probably live in jmetrics.cpp along with the MetricSink base class.  (It would avoid including it in multiple projects.)

ghalliday commented on system/metrics/sinks/log/logSink.cpp:
```diff
@@ -0,0 +1,45 @@
+
+void LogMetricSink::writeLogEntry(const std::shared_ptr<IMetric> &pMetric)
+{
+    LOG(MCmetrics, "name=%s,value=%ld,description=%s", pMetric->queryName().c_str(), pMetric->queryValue(), pMetric->queryDescription().c_str());
```
I don't think you will want the description being logged each time.

rpastrana replied:
I agree @ghalliday it seems wasteful and was going to make the same comment, I didn't make the comment because I know Prometheus seems to support reporting the description, but it would be good to know why @kenrowland decided to report it.

ghalliday commented on system/metrics/sinks/log/logSink.cpp:
```diff
@@ -0,0 +1,45 @@
+
+void LogMetricSink::writeLogEntry(const std::shared_ptr<IMetric> &pMetric)
+{
+    LOG(MCmetrics, "name=%s,value=%ld,description=%s", pMetric->queryName().c_str(), pMetric->queryValue(), pMetric->queryDescription().c_str());
```
%ld isn't the correct format specifier for a 64bit int.  i think %" I64F "d  or %lld.

ghalliday commented on system/metrics/sinks/log/logSink.hpp:
```diff
@@ -0,0 +1,42 @@
+#define LOGSINK_API DECL_IMPORT
+#endif
+
+class LOGSINK_API LogMetricSink : public PeriodicMetricSink
```
minor: This class could be in the c++ file.  In fact this file could probably be removed with everything in the c++ file.  No need for the LOGSINK_API on the class if it is in the c++ file because it can't be used outside the dll.

ghalliday commented on system/metrics/testing/metrictest.cpp:
```diff
@@ -49,13 +49,13 @@ const char *localConfigYml = R"!!(roxie:
     sinks:
         sink:
-          - type: filesink
+          - type: log
```
shouldn't this be filesink still?

ghalliday reviewed (commented):
@kenrowland a few minor comments, but generally looks good

HPCCSmoketest commented:
Process of PR-14777, label: HPCC-25630 starts now.
The reason of this test is: Forced to re-test.
Commit ID: e9f59159fbf4ef6206ba01175f8d826d5974c539
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-129.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-129.ca-central-1.compute.internal
Sha: e9f59159fbf4ef6206ba01175f8d826d5974c539
Build: success
Milestone:Install hpccsystems-platform-community_8.0.3-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |158 |158 |0 |0 |0 |58 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |25 sec  (00:00:25) |
|setup (thor) |9 |9 |0 |40 sec  (00:00:40) |
|setup (roxie) |19 |19 |0 |23 sec  (00:00:23) |
|test (hthor) |965 |965 |0 |789 sec  (00:13:09) |
|test (thor) |877 |877 |0 |1192 sec  (00:19:52) |
|test (roxie) |1038 |1037 |1 |1311 sec  (00:21:51) |


[Errors:](https://gist.github.com/ed0aa8de33cd5c23d4312c16e1b73eef)
- test
  - roxie
    - 998\. [workflow_9c (parallel=false)](https://gist.github.com/HPCCSmoketest/ed0aa8de33cd5c23d4312c16e1b73eef#file-workflow_9c-parallel-false-w20210406-190257-log), 

HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  598 sec (00:09:58) |  107 sec (00:01:47) |   21 sec (00:00:21) |   16 sec (00:00:16) | 1629 sec (00:27:09) |   18 sec (00:00:18) | 2397 sec (00:39:57) |

AttilaVamos commented:
@kenrowland I don't think the workflow_9c (parallel=false) failure is related to this PR.

ghalliday commented on system/metrics/sinks/file/fileSink.cpp:
```diff
@@ -25,82 +25,38 @@ extern "C" MetricSink* getSinkInstance(const char *name, const IPropertyTree *pS
+void FileMetricSink::writeMeasurementToFile(const std::string &metricName, uint64_t value, const std::string &metricDescription) const
 {
-    fprintf(fhandle, "  %s -> %d, %s\n", metricName.c_str(), value, metricDescription.c_str());
+    fprintf(fhandle, "  %s -> %" PRIu64 ", %s\n", metricName.c_str(), value, metricDescription.c_str());
```
Unfortunately PRIu64 is a C99 extension, and not part of c++ (it is a bit of a mess).

ghalliday commented on system/metrics/sinks/file/fileSink.hpp:
```diff
@@ -31,27 +27,21 @@ using namespace hpccMetrics;
+    ~FileMetricSink() override = default;
+
+protected:
+    void prepareToStartCollecting() override;
```
minor: As a general style, we always include the virtual keyword - even when it isn't needed.  Primarily to improve comprehension.

ghalliday commented on system/metrics/sinks/log/logSink.cpp:
```diff
@@ -0,0 +1,46 @@
+
+void LogMetricSink::writeLogEntry(const std::shared_ptr<IMetric> &pMetric)
+{
+    LOG(MCmetrics, "name=%s,value=%" PRIu64, pMetric->queryName().c_str(), pMetric->queryValue());
```
same comment about PRIu64

ghalliday reviewed (commented):
@kenrowland very close.  A few last comments.

HPCCSmoketest commented:
Process of PR-14777, label: HPCC-25630 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 024665c01bf6b973da7e97c489d7b82afa4f5a8d
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-6.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-6.ca-central-1.compute.internal
Sha: 024665c01bf6b973da7e97c489d7b82afa4f5a8d
Number of Error(s): 8
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
/home/centos/smoketest/PR-14777/HPCC-Platform/esp/platform/txsummary.cpp:29:46: error: \xe2\x80\x98createMetric\xe2\x80\x99 is not a member of \xe2\x80\x98hpccMetrics\xe2\x80\x99
/home/centos/smoketest/PR-14777/HPCC-Platform/esp/platform/txsummary.cpp:29:85: error: expected primary-expression before \xe2\x80\x98>\xe2\x80\x99 token
make[2]: *** [esp/protocols/http/CMakeFiles/esphttp.dir/build.make:336: esp/protocols/http/CMakeFiles/esphttp.dir/__/__/platform/txsummary.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:10669: esp/protocols/http/CMakeFiles/esphttp.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2021-04-07_19-18-14
time logs started
Prep time   :     9 sec (00:00:09)
Build time  :   452 sec (00:07:32)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   461 sec (00:07:41)
time logs end

[Build error:](https://gist.github.com/a44c6d69f1fd6742f039a35388f61094)
Build log [Build log](https://gist.github.com/HPCCSmoketest/a44c6d69f1fd6742f039a35388f61094#file-result-21-04-07-19-10-21-log)

HPCCSmoketest commented:
Process of PR-14777, label: HPCC-25630 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 8713010b348cd26736f92ae199699d29a4a7a2ac
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-169.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-169.ca-central-1.compute.internal
Sha: 8713010b348cd26736f92ae199699d29a4a7a2ac
Build: success
Milestone:Install hpccsystems-platform-community_8.0.3-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |158 |158 |0 |0 |0 |59 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |28 sec  (00:00:28) |
|setup (thor) |9 |9 |0 |41 sec  (00:00:41) |
|setup (roxie) |19 |19 |0 |23 sec  (00:00:23) |
|test (hthor) |965 |965 |0 |828 sec  (00:13:48) |
|test (thor) |877 |877 |0 |1174 sec  (00:19:34) |
|test (roxie) |1038 |1038 |0 |875 sec  (00:14:35) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  604 sec (00:10:04) |  113 sec (00:01:53) |   21 sec (00:00:21) |   18 sec (00:00:18) | 1497 sec (00:24:57) |   18 sec (00:00:18) | 2279 sec (00:37:59) |

kenrowland reviewed (commented):
@ghalliday @rpastrana Please take a look at the latest changes. 

Also, note that I changed the name of the templated function that creates a metric and adds it to the reporter. The new name is more descriptive. It required a change to the TxSummary file. Better change just one than multiple.

ghalliday commented on system/jlib/jlog.hpp:
```diff
@@ -849,6 +849,7 @@ constexpr LogMsgCategory MCdebugInfo(MSGAUD_programmer, MSGCLS_information, Debu
 constexpr LogMsgCategory MCauditInfo(MSGAUD_audit, MSGCLS_information, AudMsgThreshold);
 constexpr LogMsgCategory MCstats(MSGAUD_operator, MSGCLS_progress, ProgressMsgThreshold);
 constexpr LogMsgCategory MCoperatorInfo(MSGAUD_operator, MSGCLS_information, InfoMsgThreshold);
+constexpr LogMsgCategory MCmetrics(MSGAUD_operator, MSGCLS_metric, CriticalMsgThreshold);
```
minor:  It doesn't make too much difference... I'm not sure what detail level it should be, but critical seems wrong.  ErrMsgThreshold might be more appropriate.

kenrowland replied:
Changed to ErrMsgThreshold per suggestion. I used critical prior to ensure that metrics were logged regardless.

ghalliday commented on system/metrics/sinks/log/logSink.cpp:
```diff
@@ -0,0 +1,47 @@
+
+void LogMetricSink::writeLogEntry(const std::shared_ptr<IMetric> &pMetric)
+{
+//    LOG(MCmetrics, "name=%s,value=%" PRIu64, pMetric->queryName().c_str(), pMetric->queryValue());
```
trivial: Clean up the commented out lines.

kenrowland replied:
Removed.

ghalliday commented on system/metrics/testing/metrictest.cpp:
```diff
@@ -98,7 +111,7 @@ int main(int argc, char *argv[])
 
-    pQueueSizeMetric = std::make_shared<GaugeMetric>("queuesize", "request queue size");
+    pQueueSizeMetric = createMetricAndAddToReporter<GaugeMetric>("queuesize", "request queue size");
     myReporter.addMetric(pQueueSizeMetric);
```
I think this should report an error for a duplicate metric.  It would be a good unit test, but I don't think it is intended.

kenrowland replied:
The framework should throw an error if a duplicate named metric is added. However, this was not the intent of the change. I was adding a use of the function template that creates and adds a metric. Removed the extra addMetric call.

ghalliday reviewed (commented):
@kenrowland a few very small comments.  Please can you fix and also squash the commits.  Hopefully it will be ready for merging.

kenrowland reviewed (commented):
Final changes.

HPCCSmoketest commented:
Process of PR-14777, label: HPCC-25630 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 4fcc57c77910f62b594a26823f383cd816d07647
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-204.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-204.ca-central-1.compute.internal
Sha: 4fcc57c77910f62b594a26823f383cd816d07647
Build: success
Milestone:Install hpccsystems-platform-community_8.0.5-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |158 |158 |0 |0 |0 |57 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |25 sec  (00:00:25) |
|setup (thor) |9 |9 |0 |41 sec  (00:00:41) |
|setup (roxie) |19 |19 |0 |23 sec  (00:00:23) |
|test (hthor) |967 |967 |0 |828 sec  (00:13:48) |
|test (thor) |879 |879 |0 |1185 sec  (00:19:45) |
|test (roxie) |1038 |1038 |0 |885 sec  (00:14:45) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  601 sec (00:10:01) |  108 sec (00:01:48) |   21 sec (00:00:21) |   16 sec (00:00:16) | 1504 sec (00:25:04) |   18 sec (00:00:18) | 2277 sec (00:37:57) |
